### ISSUE TITLE
`khal list week` crashes 

### ISSUE THREAD

**Issue by untitaker** [pimutils/khal#490]
Since https://github.com/pimutils/khal/commit/48020f06560f3ffd0d7845d24b6464ad8f7dd3c0 `khal list week` crashes like this:

	Traceback (most recent call last):
	  File "/home/untitaker/.local/bin/khal", line 9, in <module>
		load_entry_point('khal', 'console_scripts', 'khal')()
	  File "/home/untitaker/projects/click/click/core.py", line 716, in __call__
		return self.main(*args, **kwargs)
	  File "/home/untitaker/projects/click/click/core.py", line 696, in main
		rv = self.invoke(ctx)
	  File "/home/untitaker/projects/click/click/core.py", line 1065, in invoke
		return _process_result(sub_ctx.command.invoke(sub_ctx))
	  File "/home/untitaker/projects/click/click/core.py", line 892, in invoke
		return ctx.invoke(self.callback, **ctx.params)
	  File "/home/untitaker/projects/click/click/core.py", line 534, in invoke
		return callback(*args, **kwargs)
	  File "/home/untitaker/projects/click/click/decorators.py", line 17, in new_func
		return f(get_current_context(), *args, **kwargs)
	  File "/home/untitaker/projects/khal/khal/cli.py", line 315, in klist
		env={"calendars": ctx.obj['conf']['calendars']}
	  File "/home/untitaker/projects/khal/khal/controllers.py", line 262, in khal_list
		start, end = start_end_from_daterange(daterange, conf['locale'], td)
	  File "/home/untitaker/projects/khal/khal/controllers.py", line 144, in start_end_from_daterange
		daterange, locale, default_timedelta=default_timedelta)
	  File "/home/untitaker/projects/khal/khal/aux.py", line 350, in guessrangefstr
		if len(end) == 0:
	TypeError: object of type 'datetime.datetime' has no len()

I feel like it is time to implement #169 with the `parsedatetime`.

**Comment by geier** [comment]
I don't think that parsedatetime support timeranges, only date(time)s
and timedeltas, but not combined.

But I agree, this is an idiotic bug that the test suite really should
have cought (and was introduced by my recent refactoring in
`start_end_from_daterange`).

Quoting Markus Unterwaditzer (2016-08-29 12:35:15)

**Comment by untitaker** [comment]
A timerange is just a start and an end date. You can also write "in two weeks" and it will calculate a date from this string.

On 29 August 2016 13:02:50 CEST, Christian Geier <notifications@github.com> wrote:
>
>I don't think that parsedatetime support timeranges, only date(time)s
>and timedeltas, but not combined.
>
>But I agree, this is an idiotic bug that the test suite really should
>have cought (and was introduced by my recent refactoring in
>`start_end_from_daterange`).
>
>Quoting Markus Unterwaditzer (2016-08-29 12:35:15)
>> Since
>https://github.com/pimutils/khal/commit/48020f06560f3ffd0d7845d24b6464ad8f7dd3c0
>`khal list week` crashes like this:
>> 
>>         Traceback (most recent call last):
>>           File "/home/untitaker/.local/bin/khal", line 9, in <module>
>>                 load_entry_point('khal', 'console_scripts', 'khal')()
>>           File "/home/untitaker/projects/click/click/core.py", line
>716, in __call__
>>                 return self.main(*args, **kwargs)
>>           File "/home/untitaker/projects/click/click/core.py", line
>696, in main
>>                 rv = self.invoke(ctx)
>>           File "/home/untitaker/projects/click/click/core.py", line
>1065, in invoke
>>                 return
>_process_result(sub_ctx.command.invoke(sub_ctx))
>>           File "/home/untitaker/projects/click/click/core.py", line
>892, in invoke
>>                 return ctx.invoke(self.callback, **ctx.params)
>>           File "/home/untitaker/projects/click/click/core.py", line
>534, in invoke
>>                 return callback(*args, **kwargs)
>>           File "/home/untitaker/projects/click/click/decorators.py",
>line 17, in new_func
>>                 return f(get_current_context(), *args, **kwargs)
>>           File "/home/untitaker/projects/khal/khal/cli.py", line 315,
>in klist
>>                 env={"calendars": ctx.obj['conf']['calendars']}
>>           File "/home/untitaker/projects/khal/khal/controllers.py",
>line 262, in khal_list
>>                 start, end = start_end_from_daterange(daterange,
>conf['locale'], td)
>>           File "/home/untitaker/projects/khal/khal/controllers.py",
>line 144, in start_end_from_daterange
>>                 daterange, locale,
>default_timedelta=default_timedelta)
>>           File "/home/untitaker/projects/khal/khal/aux.py", line 350,
>in guessrangefstr
>>                 if len(end) == 0:
>>         TypeError: object of type 'datetime.datetime' has no len()
>> 
>> I feel like it is time to implement #169 with the `parsedatetime`.
>> 
>> 
>> -- 
>> You are receiving this because you are subscribed to this thread.
>> Reply to this email directly or view it on GitHub:
>> https://github.com/pimutils/khal/issues/490
>
>
>-- 
>You are receiving this because you authored the thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/pimutils/khal/issues/490#issuecomment-243094769

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by geier** [comment]
no, it isn't, it can also be a datetime and a time.
But in fact, parsedatetime supports that: Calendar.evalRanges

Quoting Markus Unterwaditzer (2016-08-29 13:38:13)
> A timerange is just a start and an end date. You can also write "in two weeks" and it will calculate a date from this string.
>

**Comment by geier** [comment]
I'm less with impressed with parsedatetime's way to deal with invalid input: it never returns an error, but always something that looks valid:

```python
import parsedatetime as pdt

cal = pdt.Calendar()
print(cal.evalRanges('2011-7-8 11:00-14:00'))  # works nicely
print(cal.evalRanges('2011-7-8 11:00 14:00'))  # returns twice the right date with current time
print(cal.evalRanges('2011-2-29'))  # returns (twice) 29.8.2016!
```

**Comment by untitaker** [comment]
It can definetly be done with just the datetime parsing API. Let me prepare a pr

On 29 August 2016 14:04:22 CEST, Christian Geier <notifications@github.com> wrote:
>I'm less with impressed with parsedatetime's way to deal with invalid
>input: it never returns an error, but always something that looks
>valid:
>
>```python
>import parsedatetime as pdt
>
>cal = pdt.Calendar()
>print(cal.evalRanges('2011-7-8 11:00-14:00'))  # works nicely
>print(cal.evalRanges('2011-7-8 11:00 14:00'))  # returns twice the
>right date with current time
>print(cal.evalRanges('2011-2-29'))  # returns (twice) 29.8.2016!
>```
>
>-- 
>You are receiving this because you authored the thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/pimutils/khal/issues/490#issuecomment-243105277

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by geier** [comment]
Quoting Markus Unterwaditzer (2016-08-29 14:10:16)
> It can definetly be done with just the datetime parsing API. Let me prepare a pr
see my edit above, parsedatetime does show if its invalid or not...


>

**Comment by untitaker** [comment]
Ah ok. I was thinking of sth like this: https://github.com/pimutils/khal/pull/491

**Comment by untitaker** [comment]
Thanks!

On 30 August 2016 01:44:04 CEST, Christian Geier <notifications@github.com> wrote:
>Closed #490 via #494.
>
>-- 
>You are receiving this because you authored the thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/pimutils/khal/issues/490#event-771445712

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.
